### PR TITLE
php imap_open RCE

### PR DESCRIPTION
**Pull request by h00die**
Saw this: https://web.archive.org/web/20181118213536/https://antichat.com/threads/463395

PHP's `imap_open` call on Debian based systems can potentially be passed a parameter that is given to `rsh` which is mapped to `SSH`, which allows for RCE.

**I only coded based on the forum, I didn't discover this exploit or conribute to it in any way, all credit goes to twoster**

## Apps Reported Vulnerable:

- [ ] ~~instantcms~~ (https://github.com/instantsoft/icms2/search?q=imap_open&unscoped_q=imap_open)
  - Doesn't seem to be any reference to `imap_open` in source code 
- [ ] ~~HostCMS~~(https://github.com/HostCMS/hostcms6.free/blob/1d54bdefcd6c28918cb09dd7cb6aea404879b93f/modules/core/mail/imap.php#L140)
  - After clicking through all of HostCMS, I'm not seeing where the admin can enter the IMAP info. I believe it has the core plumbing there, simply for use by add-on modules. I'm going to abandon it for now.
- [x] e107 v2 (https://github.com/e107inc/e107/blob/7570b7ce4e17c03e9759c90889db8e750d566e53/e107_handlers/pop_bounce_handler.php#L83)
- [x] prestashop
- [x] SuiteCRM
- [ ] SugarCRM
  - no more free edition to try.  If anyone has a server running SugarCRM that will let me play with, let me know!

## Additional Apps Possible Vulnerable:

- [ ] osTicket: https://github.com/osTicket/osTicket-1.7/blob/4b94d0a2cd98d32d6bb6f63523b2db13b936e026/include/class.mailfetch.php#L138
- [ ] most likely a bunch of drupal plugins
- [ ] most likely a bunch of wordpress plugins (the most popular with 20k install when searching for 'imap' was Postie, but i can't get it to save the config, opened a help ticket with them)
- [ ] faveo-helpdesk: https://github.com/ladybirdweb/faveo-helpdesk/blob/01ba284021b1fa8e89c0e184b3560321150cb910/vendor/tedivm/fetch/src/Fetch/Server.php#L291
- [ ] yetiforceCRM; https://github.com/YetiForceCompany/YetiForceCRM/blob/68bb62fe638cddb6c3996ae3989fe1d6a96addca/modules/OSSMail/models/Record.php#L151
- [ ] postfixadmin: https://github.com/postfixadmin/postfixadmin/blob/b1795ab596e2d87cc172e698ba1f536095d5e777/model/MailboxHandler.php#L682

This is fine to land now, its VERY time consuming to install the CMS, figure out where the vuln is, write the exploit with all the darn nuances of every CMS and no core libraries (not that i think any should be added).  When I get time i'll add them to here until it lands.  After that i'll just make new PRs.

@todb-r7 I didn't see a CVE, but maybe you have more insight?  According to https://danwin1210.me this was leaked on Nov 14th.

## Verification

See docs

- [ ] Verify docs are in good working order.

### CODE REVIEW

File: modules/exploits/linux/http/php_imap_open_rce.rb [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      cookie = res.get_cookies
+      print_good('Login Success, loading admin dashboard to pull tokens')
+      res = send_request_cgi({'uri' => uri, 'cookie' => cookie})
+      #print_status(res.code.to_s)
+      #if res.code == 301 || res.code == 302
+      #  print_status("#{res.code} -> #{res.headers['Location']}")
+      #  cookie = res.get_cookies
+      #  uri = res.headers['Location']
+      #  res = send_request_cgi({'uri' => uri, 'cookie' => cookie})
+      #end
```
Comment (Green-m) [inline modules/exploits/linux/http/php_imap_open_rce.rb]: Should these comment code be deleted?

File: modules/exploits/linux/http/php_imap_open_rce.rb [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      res = send_request_cgi({
+        'uri' => normalize_uri(target_uri.path, 'index.php'),
+        'cookie' => cookie,
+        'vars_get' => {
+          'controller' => 'AdminCustomerThreads',
+          'token' => token
+        }
+      })
+
+      /form method="post" action="index\.php\?controller=AdminCustomerThreads&amp;token=(?<token>\w{32})/ =~ res.body
```
Comment (Green-m) [inline modules/exploits/linux/http/php_imap_open_rce.rb]: This way of using regular match seems to confuse me, is it better than 
`res.body =~/form method="post" action="index\.php\?controller=AdminCustomerThreads&amp;token=(?<token>\w{32})/ ` ?

**Comment by bcoles** [comment]
Jenkins test this please

File: modules/exploits/linux/http/php_imap_open_rce.rb (hunk 1 above)
Comment (h00die) [reply modules/exploits/linux/http/php_imap_open_rce.rb]: Yup, deleted now.  Prestashop is an odd product so my debugging was extensive.

**Comment by h00die** [comment]
Added suiteCRM today.  As description update says, this is OK to land now.  New PRs can be added as people find time to write in the other systems.

**Comment by h00die** [comment]
e107 is now exploitable.  Note it is a special case where `$IFS$()` was getting encoded, however `"\t"` (not `'\t'`) will bypass the filters and execute.

**Comment by syrius01** [comment]
Great share there @h00die ! I was wondering if there are some 3rd party softwares you've identified that uses php imap_open with no authentication required? This would be very critical.

Thanks

**Comment by h00die** [comment]
There's 108,000 lines of PHP code which reference "imap_open(": https://github.com/search?q=%22imap_open%28%22&type=Code
I have simply been coding what was already identified and have not been looking for new sources.  Feel free to fork this PR and submit new exploitable things.  I'd like to HIGHLY encourage that.

**Comment by h00die** [comment]
After clicking through all of HostCMS, I'm not seeing where the admin can enter the IMAP info.  I believe it has the core plumbing there, simply for use by add-on modules.  I'm going to abandon it for now.

**Comment by h00die** [comment]
@syrius01 I created a new target "custom" (no docs yet) which can be used similar to `web_delivery` for exploiting custom targets.  I believe I've found a wordpress plugin that i'll be adding shortly, but the list is endless.

**Comment by h00die** [comment]
Added "Custom" target, and finished documenting it, so you can now get the string to use, and have a listener for attempts to exploit custom apps, similar to `web_delivery`.

I'm most likely done now.  I updated the description to incorporate all the additional findings.

**Comment by busterb** [comment]
Wow, this is nice! I'm going to go ahead and get it into master so folks can take advantage of it. Sounds like plenty to add over time.

**Comment by todb-r7** [comment]
Thanks @h00die ! I'm alerting the [DWF](https://docs.google.com/forms/d/e/1FAIpQLSddsMgF0JuiMlcURdmVuIdGtEgDNKXhjNy6ZoRwGrSb6Ty60g/formResponse) right now, they should kick out a CVE for you/twoster.

**Comment by busterb** [comment]
## Release Notes

This adds an exploit for PHP's imap_open call, which on Debian-based systems can potentially be passed a parameter that is given to rsh which is mapped to SSH, allowing for remote code execution.

**Comment by todb-r7** [comment]
Based on your notes, @h00die , I am presuming you were testing the very latest PHP, PHP 7.3.0RC6 , or at least something close enough that this functionality is quite likely to be present there. In any event, the 7.3 release branch doesn't mention imap_open at all in the [UPGRADING](https://github.com/php/php-src/blob/php-7.3.0RC6/UPGRADING) notes, so I'm assuming it's untouched.

**Comment by h00die** [comment]
@todb-r7 I think I only tested 7.0 but will check a little later and confirm the version and try the latest and post back

**Comment by todb-r7** [comment]
And @busterb just found this CVE listing:

https://vulmon.com/vulnerabilitydetails?qid=CVE-2018-19518

Which says the bug is in U of W  IMAP Toolkit, whatever that is, not PHP, which seems incorrect. Hmm. Will let DWF sort it out.

**Comment by busterb** [comment]
Also original PHP bug report from months ago: https://bugs.php.net/bug.php?id=76428

**Comment by h00die** [comment]
@todb-r7 that reference the original code from @Bo0oM so we can go with that.  I'll put in a new PR to update this code w/ the CVE you found (nice find!)

**Comment by todb-r7** [comment]
Got a CVE from DWF: CVE-2018-1000859

I can edit up the module nowish to mention it.

**Comment by h00die** [comment]
Thanks @todb-r7 that would be great. I think I may have a new target soon but can't start till my laptop gets fixed (hoping to get parts in today).

**Pull request merged**